File: server/lyrics.py
from abc import ABC, abstractmethod
import base64
import requests
from pykakasi import Kakasi
from unicodeblock import blocks as unicodeBlock
from xmltodict import parse as parseXml

TERM_FURI_REPLACEMENTS = {
    "日": "ひ",
    "君": "きみ",
    "色": "いろ",
    "瞑": "つぶ",
    "足": "あし",
    "角": "かど"
}

class Lyrics(ABC):

    def __init__(self, id: str):
        self.id = id

    def getId(self) -> str:
        return self.id

    def isTimed(self) -> bool:
        return isinstance(self, TimedLyrics)

    def getSource(self) -> str:
        return self.getId().split("/")[0]

    def getText(self) -> str:
        if isinstance(self, StaticLyrics):
            return self.text
        else:
            assert(isinstance(self, TimedLyrics))
            ret = ""
            for line in self.lines:
                ret += line.text + "\n"
            return ret

    def _getFuriganaData(self):
        def hasKanjiAndHiragana(string: str):
            has_kanji = False
            has_hiragana = False

            for char in string:
                block = unicodeBlock.of(char)
                if block == "HIRAGANA":
                    if has_kanji:
                        return True
                    has_hiragana = True
                elif block == "CJK_UNIFIED_IDEOGRAPHS":
                    if has_hiragana:
                        return True
                    has_kanji = True

        def isKanji(char: str):
            return unicodeBlock.of(char) == "CJK_UNIFIED_IDEOGRAPHS"

        def hasKanji(string: str):
            for char in string:
                if isKanji(char):
                    return True
            return False

        def trimOkurigana(original: str, furigana: str) -> list:
            if (hasKanjiAndHiragana(original)):
                trim_amount = 0
                for i in range(1, len(furigana) + 1):
                    if isKanji(original[len(original) -i]) or original[len(original) -i] != furigana[len(furigana) - i]:
                        trim_amount = i - 1
                        break

                if trim_amount != 0:
                    return [[
                        original[:len(original) - trim_amount],
                        furigana[:len(furigana) - trim_amount]
                    ], [original[len(original) - trim_amount:len(original)]]]

            return [[original, furigana]]

        def getKey(term, key: str) -> str:
            return term[key].replace("\\n", "\n").replace("\\r", "\r")

        kakasi = Kakasi()
        ret = []

        for line in [line.text for line in self.lines] if isinstance(self, TimedLyrics) else self.getText().split("\n"):

            line_data = []
            ret.append(line_data)

            for term in kakasi.convert(line.replace("\n", "\\n").replace("\r", "\\r")):

                orig = getKey(term, "orig")
                hira = getKey(term, "hira")

                if orig != hira and hasKanji(orig):
                    terms = trimOkurigana(orig, hira)
                    for i in range(len(terms)):
                        if len(terms[i]) == 0:
                            continue
                        replacement = TERM_FURI_REPLACEMENTS.get(terms[i][0])
                        if replacement is not None:
                            terms[i][1] = replacement

                    line_data += terms
                else:
                    line_data.append([orig])

        return ret

    @abstractmethod
    def getWithFurigana(self) -> list:
        raise NotImplementedError

class StaticLyrics(Lyrics):
    def __init__(self, id: str, text: str):
        super().__init__(id)
        self.text = text

    def getWithFurigana(self) -> list:
        ret = []

        for line in self._getFuriganaData():
            terms = []
            for term in line:
                if term[0] == "" or (term[0] == "\r" and len(line) > 1):
                    continue
                terms.append({"subterms": [{"text": term[0], "furi": term[1] if len(term) > 1 else None}]})
            if len(terms) > 0:
                ret.append(terms)

        return ret

class TimedLyrics(Lyrics):

    def __init__(self, id: str, xml_data: str):
        super().__init__(id)
        prev_line = None
        index = 0

        self.lines: list[TimedLyrics.Line] = []
        self.first_word: TimedLyrics.Word | None = None

        for line_data in parseXml(xml_data)["wsy"]["line"]:
            line = TimedLyrics.Line(line_data, index)

            self.lines.append(line)
            index += len(line.text)
            if prev_line != None:
                prev_line.next_line = line
                prev_line.words[-1].next_word = line.words[0]
                line.words[0].prev_word = prev_line.words[-1]

            line.prev_line = prev_line
            prev_line = line

            if self.first_word == None:
                self.first_word = line.words[0]

    class Word:

        def __init__(self, data: dict, index: int):
            self.text = data["wordstring"]
            self.start_time = int(data["starttime"]) / 1000
            self.end_time = int(data["endtime"]) / 1000
            self.next_word: TimedLyrics.Word | None = None
            self.prev_word: TimedLyrics.Word | None = None
            self.index = index

    class Line:

        def __init__(self, data: dict, _index: int):
            self.text: str = data["linestring"]
            self.is_space = self.text == None
            self.words: list[TimedLyrics.Word] = []
            self.next_line: TimedLyrics.Line | None = None
            self.prev_line: TimedLyrics.Line | None = None

            self.lines: list = []
            self.first_word: TimedLyrics.Word | None = None

            if self.is_space:
                self.words.append(TimedLyrics.Word({
                    "wordstring": "\n",
                    "starttime": -1,
                    "endtime": -1
                }, _index))
                self.text = "\n"
            else:
                index = _index
                prev_word: TimedLyrics.Word | None = None
                for word_data in data["word"] if isinstance(data["word"], list) else [data["word"]]:
                    if word_data["wordstring"] is None:
                        continue

                    word = TimedLyrics.Word(word_data, index)
                    if len(word.text) > 0:
                        self.words.append(word)
                        index += len(word.text)
                        if prev_word != None:
                            prev_word.next_word = word

                        word.prev_word = prev_word
                        prev_word = word

    def getWithFurigana(self) -> list:
        ret = []
        furigana = self._getFuriganaData()

        for line_i in range(len(self.lines)):

            line_entry = []
            ret.append(line_entry)

            lyr_line = self.lines[line_i]
            if lyr_line.is_space:
                line_entry.append({"subterms": [{"text": "", "furi": None}], "start": -1, "end": -1})
                continue

            furi_line: list = furigana[line_i]
            borrow = 0

            for word in lyr_line.words:
                terms: list[dict[str, str | None]] = []
                i = len(word.text) - borrow
                borrow = max(0, borrow - len(word.text))

                while i > 0 and len(furi_line) > 0:
                    orig = furi_line[0][0]
                    furi = furi_line[0][1] if len(furi_line[0]) > 1 else None

                    if len(orig.strip()) == 0:
                        furi_line.pop(0)
                        continue

                    # Term fits into word
                    if len(orig) <= i:
                        i -= len(orig)
                        terms.append({"text": orig, "furi": furi})
                        furi_line.pop(0)

                    # Term doesn't fit into word
                    else:

                        # Term has no furigana, so we can disect it
                        if furi is None:
                            terms.append({"text": orig[:i], "furi": None})
                            furi_line[0][0] = orig[i:]
                        else:
                            terms.append({"text": orig, "furi": furi})
                            furi_line.pop(0)
                            borrow = len(orig) - i

                        i = 0

                if len(terms) > 0:
                    word_entry = {"subterms": terms, "start": word.start_time, "end": word.end_time}
                    line_entry.append(word_entry)

        return ret

def getLyricsData(song_id: int, lyrics_type: int) -> str | None:
    if lyrics_type <= 0 or lyrics_type > 3:
        raise RuntimeError()

    response = requests.post(
        "https://p1.petitlyrics.com/api/GetPetitLyricsData.php",
        data={
            "key_lyricsId": song_id,
            "lyricsType": lyrics_type,
            "terminalType": "10",
            "clientAppId": "on354007"
        })

    response.raise_for_status()

    try:
        lyrics_data: str = parseXml(response.text)["response"]["songs"]["song"]["lyricsData"]
        return base64.b64decode(lyrics_data).decode("utf-8")
    except (KeyError, UnicodeDecodeError):
        return None

def getLyrics(lyrics_id: str) -> Lyrics | None:

    split = lyrics_id.split(":")
    if len(split) != 2:
        return None

    method = split[0]
    id = split[1]

    match (method):
        case "ptl":
            data = None
            try:
                song_id = int(id)
            except ValueError:
                return None

            for i in range(3, 0, -1):
                data = getLyricsData(song_id, i)
                if data is not None:
                    break

            if data is None:
                return None

            return TimedLyrics(lyrics_id, data) if data.startswith("<wsy>") and data.endswith("</wsy>") else StaticLyrics(lyrics_id, data)
        case _:
            return None

    # data = getLyricsData(song_id, 1)
    # if data is not None:
    #     return StaticLyrics(data)

    # return None

def searchForLyrics(title: str, artist: str | None = None) -> list[dict[str, str]]:

    params = {"title": title}
    if artist is not None:
        params["artist"] = artist

    # Replace all whitespace with the standard character
    for key in params:
        new = ""
        for char in params[key]:
            if char.isspace():
                new += " "
            else:
                new += char
        params[key] = new

    result_start = "<a href=\"/lyrics/"
    result_end = "</a>"

    def getResults():
        response = requests.get("https://petitlyrics.com/search_lyrics", params=params)
        response.raise_for_status()
        data = response.text

        ret = []
        result = {}

        start = data.find(result_start)
        while start != -1:
            href = data[start + len(result_start) : data.find("\"", start + len(result_start) + 1)]
            end = data.find(result_end, start + len(result_start))

            try:
                # If first char is an int, this is the start of a new result
                result_id = int(href)

                if "id" in result:
                    ret.append(result)
                    result = {}
                else:
                    result.clear()

                parsed = parseXml(data[start : end + len(result_end)])
                result["id"] = f"ptl:{result_id}"
                result["name"] = parsed["a"]["span"]["#text"]

            except ValueError:
                split = href.split("/")

                if split[0] in ("artist", "album") and len(split) > 1 and len(split[1]) != 0:
                    parsed = parseXml(data[start : end + len(result_end)])

                    result[f"{split[0]}_id"] = split[1]
                    result[f"{split[0]}_name"] = parsed["a"]["span"]["#text"]

            start = data.find(result_start, end + len(result_end))

        if "id" in result:
            ret.append(result)

        return ret

    ret = getResults()
    if len(ret) == 0 and artist is not None:
        params.pop("artist")
        return getResults()
    return ret

File: shared/src/commonMain/kotlin/com/toasterofbread/spmp/api/Lyrics.kt
package com.toasterofbread.spmp.api

import com.atilika.kuromoji.ipadic.Tokenizer
import com.toasterofbread.spmp.model.mediaitem.Song
import com.toasterofbread.spmp.model.SongLyrics
import com.toasterofbread.utils.hasKanjiAndHiragana
import com.toasterofbread.utils.isKanji
import com.toasterofbread.utils.isJP
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.jsoup.Jsoup
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.nio.channels.ClosedByInterruptException
import java.util.*

suspend fun getSongLyrics(song: Song, data: Pair<Int, SongLyrics.Source>?): Result<SongLyrics> {
    if (song.title == null) {
        return Result.failure(RuntimeException("Song has no title"))
    }

    val result: Result<SongLyrics> =
        if (data != null) getLyrics(data.first, data.second)
        else {
            searchForLyrics(song.title!!, song.artist?.title).fold(
                { results ->
                    if (results.isEmpty()) Result.failure(RuntimeException("No lyrics found"))
                    else getLyrics(results.first().id, results.first().source)
                },
                {
                    Result.failure(it)
                }
            )
        }

    result.getOrNull()?.also { lyrics ->
        song.song_reg_entry.updateLyrics(lyrics.id, lyrics.source)
    }

    return result
}

fun getLyricsData(lyrics_id: Int, sync_type: SongLyrics.SyncType): Result<String> {
    val body = "key_lyricsId=$lyrics_id&lyricsType=${sync_type.ordinal + 1}&terminalType=10&clientAppId=on354007".toRequestBody("application/x-www-form-urlencoded; charset=utf-8".toMediaType())
    val request = Request.Builder()
        .url("https://p1.petitlyrics.com/api/GetPetitLyricsData.php")
        .post(body)
        .build()

    val result = Api.request(request)
    if (result.isFailure) {
        return result.cast()
    }

    val START = "<lyricsData>"
    val END = "</lyricsData>"

    val xml = result.getOrThrowHere().body!!.string()
    val start = xml.indexOf(START)
    val end = xml.indexOf(END, start + START.length)

    val decoded = Base64.getDecoder().decode(xml.substring(start + START.length, end))
    return Result.success(String(decoded))
}

private fun trimOkurigana(term: SongLyrics.Term.Text): List<SongLyrics.Term.Text> {

    if (term.furi == null || !term.text.hasKanjiAndHiragana()) {
        return listOf(term)
    }

    var trim_start: Int = 0
    for (i in 0 until term.furi!!.length) {
        if (term.text[i].isKanji() || term.text[i] != term.furi!![i]) {
            trim_start = i
            break
        }
    }

    var trim_end: Int = 0
    for (i in 1 .. term.furi!!.length) {
        if (term.text[term.text.length - i].isKanji() || term.text[term.text.length - i] != term.furi!![term.furi!!.length - i]) {
            trim_end = i - 1
            break
        }
    }

    val terms: MutableList<SongLyrics.Term.Text> = mutableListOf()
    val last_term: SongLyrics.Term.Text

    if (trim_start > 0) {
        terms.add(
            SongLyrics.Term.Text(
                term.text.substring(0, trim_start),
                null
            )
        )

        last_term = SongLyrics.Term.Text(
            term.text.substring(trim_start),
            term.furi!!.substring(trim_start)
        )
    }
    else {
        last_term = term
    }

    if (trim_end > 0) {
        terms.add(
            SongLyrics.Term.Text(
                last_term.text.substring(0, last_term.text.length - trim_end),
                last_term.furi!!.substring(0, last_term.furi!!.length - trim_end)
            )
        )
        terms.add(
            SongLyrics.Term.Text(
                last_term.text.takeLast(trim_end),
                null
            )
        )
    }
    else {
        terms.add(last_term)
    }

    return terms
}

private fun mergeAndFuriganiseTerms(tokeniser: Tokenizer, terms: List<SongLyrics.Term>): List<SongLyrics.Term> {
    if (terms.isEmpty()) {
        return emptyList()
    }

    val ret: MutableList<SongLyrics.Term> = mutableListOf()
    val terms_to_process: MutableList<SongLyrics.Term> = mutableListOf()

    for (term in terms) {
        val text = term.subterms.single().text
        if (text.any { it.isJP() }) {
            terms_to_process.add(term)
        }
        else {
            ret.addAll(_mergeAndFuriganiseTerms(tokeniser, terms_to_process))
            terms_to_process.clear()
            ret.add(term)
        }
    }

    ret.addAll(_mergeAndFuriganiseTerms(tokeniser, terms_to_process))

    return ret
}

private fun _mergeAndFuriganiseTerms(tokeniser: Tokenizer, terms: List<SongLyrics.Term>): List<SongLyrics.Term> {
    if (terms.isEmpty()) {
        return emptyList()
    }

    val ret: MutableList<SongLyrics.Term> = mutableListOf()
    val line_range = terms.first().line_range!!
    val line_index = terms.first().line_index

    var terms_text: String = ""
    for (term in terms) {
        terms_text += term.subterms.single().text
    }

    val tokens = tokeniser.tokenize(terms_text)

    var current_term: Int = 0
    var term_head: Int = 0

    for (token in tokens) {
        val token_base = token.surface

        var text: String = ""
        var start: Long = Long.MAX_VALUE
        var end: Long = Long.MIN_VALUE

        while (text.length < token_base.length) {
            val term = terms[current_term]
            val subterm = term.subterms.single()
            start = minOf(start, term.start!!)
            end = maxOf(end, term.end!!)

            val needed = token_base.length - text.length
            if (needed < subterm.text.length - term_head) {
                text += subterm.text.substring(term_head, term_head + needed)
                term_head += needed
            }
            else {
                text += subterm.text
                term_head = 0
                current_term++
            }
        }

        val term = SongLyrics.Term(trimOkurigana(SongLyrics.Term.Text(text, token.reading)), line_index, start, end)
        term.line_range = line_range
        ret.add(term)
    }

    return ret
}

private fun parseStaticLyrics(data: String): List<List<SongLyrics.Term>> {
    val tokeniser = Tokenizer()

    val ret: MutableList<List<SongLyrics.Term>> = mutableListOf()

    for (line in data.split('\n')) {
        val tokens = tokeniser.tokenize(line)
        ret.add(tokens.mapNotNull { token ->
            if (token.surface.isBlank()) null
            else SongLyrics.Term(listOf(SongLyrics.Term.Text(token.surface, token.reading)), ret.size)
        })
    }

    return ret
}

private fun parseTimedLyrics(data: String): List<List<SongLyrics.Term>> {
    val parser = XmlPullParserFactory.newInstance().newPullParser()
    parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
    parser.setInput(data.reader())
    parser.nextTag()

    fun skip() {
        if (parser.eventType != XmlPullParser.START_TAG) {
            throw IllegalStateException()
        }
        var depth = 1
        while (depth != 0) {
            when (parser.next()) {
                XmlPullParser.END_TAG -> depth--
                XmlPullParser.START_TAG -> depth++
            }
        }
    }

    fun readText(tag: String): String {
        parser.require(XmlPullParser.START_TAG, null, tag)

        var result = ""
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }

        parser.require(XmlPullParser.END_TAG, null, tag)
        return result
    }

    fun parseTerm(line_index: Int): SongLyrics.Term? {
        parser.require(XmlPullParser.START_TAG, null, "word")

        var text: String? = null
        var start: Long? = null
        var end: Long? = null

        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }

            when (parser.name) {
                "wordstring" -> text = readText("wordstring")
                "starttime" -> start = readText("starttime").toLong()
                "endtime" -> end = readText("endtime").toLong()
                else -> skip()
            }
        }

        parser.require(XmlPullParser.END_TAG, null, "word")

        if (text!!.isEmpty()) {
            return null
        }

        return SongLyrics.Term(listOf(SongLyrics.Term.Text(text)), line_index, start!!, end!!)
    }

    val tokeniser: Tokenizer
    try {
        tokeniser = Tokenizer()
    }
    catch (e: RuntimeException) {
        if (e.cause is ClosedByInterruptException) {
            throw InterruptedException()
        }
        else {
            throw e
        }
    }

    fun parseLine(index: Int): List<SongLyrics.Term> {
        parser.require(XmlPullParser.START_TAG, null, "line")

        var line_start: Long? = null
        var line_end: Long? = null

        val terms: MutableList<SongLyrics.Term> = mutableListOf()
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }

            if (parser.name != "word") {
                skip()
                continue
            }

            val term = parseTerm(index)
            if (term != null) {
                terms.add(term)

                if (line_start == null) {
                    line_start = term.start
                    line_end = term.end
                }
                else {
                    if (term.start!! < line_start) {
                        line_start = term.start
                    }
                    if (term.end!! > line_end!!) {
                        line_end = term.end
                    }
                }
            }
        }

        if (terms.isNotEmpty()) {
            val range = line_start!! .. line_end!!
            for (term in terms) {
                term.line_range = range
            }
        }

        parser.require(XmlPullParser.END_TAG, null, "line")
        return mergeAndFuriganiseTerms(tokeniser, terms)
    }

    val ret: MutableList<List<SongLyrics.Term>> = mutableListOf()

    parser.require(XmlPullParser.START_TAG, null, "wsy")
    while (parser.next() != XmlPullParser.END_TAG) {
        if (parser.eventType != XmlPullParser.START_TAG) {
            continue
        }

        if (parser.name != "line") {
            skip()
            continue
        }

        ret.add(parseLine(ret.size))
    }

    while (ret.first().isEmpty()) {
        ret.removeFirst()
    }
    while (ret.last().isEmpty()) {
        ret.removeLast()
    }

    for (line in ret.withIndex()) {
        if (line.value.isEmpty()) {
            ret[line.index] = listOf()
        }
    }

    return ret
}

fun getLyrics(lyrics_id: Int, lyrics_source: SongLyrics.Source): Result<SongLyrics> {
    when (lyrics_source) {
        SongLyrics.Source.PETITLYRICS -> {
            for (sync_type in SongLyrics.SyncType.byPriority()) {
                val result = getLyricsData(lyrics_id, sync_type)
                if (!result.isSuccess) {
                    return result.cast()
                }

                val lyrics: List<List<SongLyrics.Term>>
                if (result.data.startsWith("<wsy>")) {
                    lyrics = parseTimedLyrics(result.data)
                }
                else {
                    lyrics = parseStaticLyrics(result.data)
                    return Result.success(SongLyrics(lyrics_id, lyrics_source, SongLyrics.SyncType.NONE, lyrics))
                }

                return Result.success(SongLyrics(lyrics_id, lyrics_source, sync_type, lyrics))
            }
        }
    }

    return Result.failure(NotImplementedError())
}

private fun concatParams(first: String, second: String): String {
    var ret = first
    for (char in second) {
        // Replace all whitespace with the standard character
        if (char.isWhitespace()) {
            ret += ' '
        }
        else {
            ret += char
        }
    }
    return ret
}

data class LyricsSearchResult(
    var id: Int,
    val source: SongLyrics.Source,
    var name: String,
    var sync_type: SongLyrics.SyncType,
    var artist_id: String?,
    var artist_name: String?,
    var album_id: String?,
    var album_name: String?
)

suspend fun searchForLyrics(title: String, artist: String?): Result<List<LyricsSearchResult>> = withContext(Dispatchers.IO) {

    val title_param = concatParams("?title=", title)
    val artist_param = if (artist != null) concatParams("&artist=", artist) else ""

    val RESULT_START = "<a href=\"/lyrics/"
    val RESULT_END = "</a>"
    val SYNC_TYPE_START = "<span class=\"lyrics-list-sync "

    fun performSearch(params: String): Result<List<LyricsSearchResult>> {
        val request = Request.Builder()
            .url("https://petitlyrics.com/search_lyrics$params")
            .header("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/114.0")
            .build()

        val result = Api.request(request)
        if (result.isFailure) {
            return result.cast()
        }

        val ret = mutableListOf<LyricsSearchResult>()

        var r_id: Int? = null
        val r_source: SongLyrics.Source = SongLyrics.Source.PETITLYRICS
        var r_name: String? = null
        var r_sync_type: SongLyrics.SyncType? = null
        var r_artist_id: String? = null
        var r_artist_name: String? = null
        var r_album_id: String? = null
        var r_album_name: String? = null

        val lines = result.getOrThrowHere().body!!.string().split('\n')
        for (element in lines) {
            val line = element.trim()

            if (!line.startsWith(RESULT_START)) {
                if (r_id != null && r_sync_type == null && line.startsWith(SYNC_TYPE_START)) {
                    val sync_type = line.substring(SYNC_TYPE_START.length, line.indexOf('"', SYNC_TYPE_START.length))
                    r_sync_type = SongLyrics.SyncType.fromKey(sync_type)
                }
                continue
            }

            val href = line.substring(RESULT_START.length, line.indexOf('"', RESULT_START.length + 1))
            val end = line.indexOf(RESULT_END, RESULT_START.length + href.length)

            // If href is an int, this is the start of a new result
            val result_id = href.toIntOrNull()
            if (result_id != null) {
                if (r_id != null) {
                    ret.add(
                        LyricsSearchResult(
                            r_id,
                            r_source,
                            r_name!!,
                            r_sync_type!!,
                            r_artist_id,
                            r_artist_name,
                            r_album_id,
                            r_album_name
                        )
                    )
                }
                else {
                    r_sync_type = null
                    r_artist_id = null
                    r_artist_name = null
                    r_album_id = null
                    r_album_name = null
                }

                r_id = result_id
                r_name = Jsoup.parse(line.substring(0, end + RESULT_END.length)).body().text()
            }
            else {
                val split = href.split('/')

                when (split[0]) {
                    "artist" -> {
                        r_artist_id = split[1]
                        r_artist_name = Jsoup.parse(line.substring(0, end + RESULT_END.length)).body().text()
                    }
                    "album" -> {
                        r_album_id = split[1]
                        r_album_name = Jsoup.parse(line.substring(0, end + RESULT_END.length)).body().text()
                    }
                }
            }
        }

        if (r_id != null) {
            ret.add(
                LyricsSearchResult(
                    r_id,
                    r_source,
                    r_name!!,
                    r_sync_type!!,
                    r_artist_id,
                    r_artist_name,
                    r_album_id,
                    r_album_name
                )
            )
        }

        return Result.success(ret)
    }

    try {
        val result = performSearch(title_param + artist_param)
        result.onSuccess { results ->
            if (results.isEmpty() && artist != null) {
                return@withContext performSearch(title_param)
            }
        }
        return@withContext result
    }
    catch (e: Throwable) {
        return@withContext Result.failure(e)
    }
}
